### DISCUSSION TITLE
Companion simulator

### DISCUSSION THREAD

**Discussion by dwheeld** [EdgeTX/edgetx#908] [Q&A] [answered]
I am trying to understand how Companion code handles the telemetry ID numbers. In telemetrysimu.cpp it sets the id numbers in the code listed below. Before it can SET the ID’s it has to know what they are. However, I have searched the entire cpp and can not find where it obtains the ID values. It doesn’t open any files, no signals and slots are associated, and no global variables. So where the heck do the values come from?

  SET_INSTANCE(rxbt_inst,     BATT_ID,                0);
  SET_INSTANCE(rssi_inst,     RSSI_ID,                24);
  SET_INSTANCE(swr_inst,      RAS_ID,                 24);
  SET_INSTANCE(a1_inst,       ADC1_ID,                0);
  SET_INSTANCE(a2_inst,       ADC2_ID,                0);
  SET_INSTANCE(a3_inst,       A3_FIRST_ID,            0);
  SET_INSTANCE(a4_inst,       A4_FIRST_ID,            0);
  SET_INSTANCE(t1_inst,       T1_FIRST_ID,            0);
  SET_INSTANCE(t2_inst,       T2_FIRST_ID,            0);
  SET_INSTANCE(rpm_inst,      RPM_FIRST_ID,           DATA_ID_RPM);
  SET_INSTANCE(fuel_inst,     FUEL_FIRST_ID,          0);
  SET_INSTANCE(fuel_qty_inst, FUEL_QTY_FIRST_ID,      0);
  SET_INSTANCE(aspd_inst,     AIR_SPEED_FIRST_ID,     0);
  SET_INSTANCE(vvspd_inst,    VARIO_FIRST_ID,         DATA_ID_VARIO);
  SET_INSTANCE(valt_inst,     ALT_FIRST_ID,           DATA_ID_VARIO);
  SET_INSTANCE(fasv_inst,     VFAS_FIRST_ID,          DATA_ID_FAS);
  SET_INSTANCE(fasc_inst,     CURR_FIRST_ID,          DATA_ID_FAS);
  SET_INSTANCE(cells_inst,    CELLS_FIRST_ID,         DATA_ID_FLVSS);
  SET_INSTANCE(gpsa_inst,     GPS_ALT_FIRST_ID,       DATA_ID_GPS);
  SET_INSTANCE(gpss_inst,     GPS_SPEED_FIRST_ID,     DATA_ID_GPS);
  SET_INSTANCE(gpsc_inst,     GPS_COURS_FIRST_ID,     DATA_ID_GPS);
  SET_INSTANCE(gpst_inst,     GPS_TIME_DATE_FIRST_ID, DATA_ID_GPS);
  SET_INSTANCE(gpsll_inst,    GPS_LONG_LATI_FIRST_ID, DATA_ID_GPS);
  SET_INSTANCE(accx_inst,     ACCX_FIRST_ID,          0);
  SET_INSTANCE(accy_inst,     ACCY_FIRST_ID,          0);
  SET_INSTANCE(accz_inst,     ACCZ_FIRST_ID,          0);

Also, where is the code that creates and updates the telemetry screen for a radiomaster tx16s in the companion simulator?

![telemetry screen](https://user-images.githubusercontent.com/7997610/137503916-6d8830e8-bf44-4cd5-9be0-d19ed8fdc2f0.jpg)

**Comment by raphaelcoeffic** [comment]
It’s a window form that you can open as a side window. Don’t remember the exact button though.

**Reply by raphaelcoeffic** [reply]
@elecpower it would be nice if you could give this gentlemen some details.

**Comment by elecpower** [comment, marked as answer]
@dwheeld  Companion Simulator is a wrapper to run a simplified version of the radio code and allow interaction. As the code is not running on a physical radio a lot of runtime radio functions have to be 'faked' or disabled.
IDs from radio/src/telemetry/frsky.h
The telemetry screen is also the radio code.

**Reply by dwheeld** [reply]
Thanks. that is the answer I was looking for. I had followed the #include "radio/src/telemetry/frsky.h" back, but didn't explore it thoroughly. That gives me something work with now. :)
